### DISCUSSION TITLE
How can people help prepare the v0.1.0 release?

### DISCUSSION THREAD

**Discussion by pieterdd** [lapce/floem#232] [General]
Having worked with Floem for a good week or so now, I see a lot of potential in it. It's fairly easy to learn compared to other native Rust GUI frameworks I've looked at, and yet it's quite flexible. I realize that as maintainers you are probably more focused on Floem's shortcomings. But for hobby projects I already prefer Floem over Qt, Flutter or Tauri. Native GUI development has long been a struggle within the Rust ecosystem. I think Floem holds the key to changing that. So by all means, give yourselves a pat on the back for everything you've already achieved.

To me as a casual user, it feels like Floem is more than mature enough for a 0.1.0 release. To entice more developers to consider native Rust GUIs and lower effort involved in building them, I'd like to publish a [Floem UI toolkit](https://github.com/pieterdd/floem-ui-kit/) to crates.io. Since crates.io does not support Git-based dependencies like Floem, I was wondering: which tasks are blockers for a 0.1.0 release? And are you accepting patches for those tasks?

**Comment by jrmoulton** [comment]
https://discord.com/channels/946858761413328946/1071098441289179267/1184899506026856508

There have been a few discussions around this in the discord. It seems the expectation is for a release after a proper/improved text widget is added. 

Pull requests are always welcome but for working on this it would probably be good to check with @presiyan-ivanov and @MinusGix to see what the current state of things is.

**Reply by pieterdd** [reply]
Could you clarify the criteria for a "proper" text widget? Which missing features would need to be added?

**Reply by jrmoulton** [reply]
I'm not actually sure what they are, I haven't worked on the text input or seen any specifics in the discord. So the others I mentioned will know better than me.

If there is a checklist of items I'd love to jump in and help with implementing them as well

**Reply by dominikwilkowski** [reply]
text improvements here: https://github.com/lapce/floem/issues/28

**Reply by presiyan-ivanov** [reply]
Its mainly about IME and multi-line support.
Multi-line is quite involved to get right,  so the plan is that lapce's own text input(which is far more sophisticated) gets added to floem instead. 

Technically the text edit functionality will be extracted to a library and will be used by floem, so you can sync with @MinusGix  on how you can help with that, if you are interested.

**Reply by pieterdd** [reply]
Thanks for the extra context. If multiline and IME support require large time investments, I'd be happy to use a release that doesn't yet include them. For my personal needs, Floem is definitely capable enough to merit an alpha release even without them.

That said, I do respect the decision to consider these blockers, and I'll take a look at the codebase to see if I have the relevant skillset to help.

**Reply by MinusGix** [reply]
See: https://github.com/lapce/lapce/pull/2909

**Comment by dominikwilkowski** [comment]
Question from me: why release 0.1.0? they are unstable and each change is considered a major breaking change so how is that different from just grabbing floem from git directly? What does a release give us here?
My personal opinion is that there are some major issues in floem (including but not exclusively multi line input fields) that have yet to mature and I'd rather personally go to v1 than to v0 releases but that's just me.

**Reply by pieterdd** [reply]
I'm not sure I should voice an opinion on this, because ultimately the best workflow for a volunteer-led project is what you and the other Floem devs enjoy the most. Since you did ask the question, I will try to answer it.

The underlying assumption we're both making, I think, is that more people will discover and try Floem in its unfinished form if there are published releases. 0.x releases may help in collecting early feedback from real-world users. That feedback can shape the v1 architecture so that Floem's developer experience will be as good as it can be. Consider too that more well-known Rust GUI frameworks aren't necessarily better than Floem. They did however run `cargo publish`, which made them more discoverable. So the most well-known libraries aren't necessarily the best ones. This may lead users to pick a library that's less polished than Floem or give up on Rust for writing GUIs altogether - just because they didn't know Floem exists. Maybe they would have even contributed to Floem if they were aware of its existence.

I understand the hesitance to release something that you feel is very rough around the edges. After all, you know Floem's limitations way better than me as a casual user. But if the README explicitly mentions that people may encounter bugs and frequent breaking API changes until v1 is established, then my expectations as a user have been set appropriately.

But again, I'm just happy that you guys are building Floem, so I just wanna let you do what you do best.

**Reply by dominikwilkowski** [reply]
FWIW I'm new to floem and have been using it in my own project for the last couple of months and been loving it. The maintainers can make this call. No strong opinions from me. Your point about discoverability if sound. Rust has such a big GUI problem I don't think it'll be a problem for floem though.

**Reply by panekj** [reply]
You can check https://github.com/lapce/floem/issues/163 why some people would like to see it as a crate/proper release instead of just git source

**Comment by Zoxc** [comment]
One requirement be to deal with the library forks Floem uses. We'd need to move to upstream and published versions of those. That would involve cleaning up the changes in those fork and upstreaming them.

I think a basic set of widgets would be helpful too. @pieterdd You may be interested in that. Feel free to stop by the Discord too if you're interesting in contributing.

**Reply by pieterdd** [reply]
Ah, good point. It hadn't occurred to me yet that Floem imports Peniko and Winit as Git dependencies. The Peniko change points to a commit that's already in their main branch, they just haven't done any releases yet. Fortunately there aren't any Git dependencies in their Cargo.toml, so it might be a matter of me talking to the maintainers. The Winit one seems a lot trickier. My dayjob starts back up again tomorrow, but I might find some time to dive into it later this week.

Regarding shipping a default set of widgets, that's a bigger topic consisting of a few subtopics I'll have to revisit when time allows.

**Reply by Zoxc** [reply]
`cosmic-text` and `vger-rs` also have forks. I suspect the latter might be replaced by [Vello](https://github.com/linebender/vello) eventually.

**Comment by pieterdd** [comment]
I noticed that the first 0.x release is live! Thank you to everyone who contributed to this early version, and I'm curious to see Floem evolve from here.
